beacon_node/beacon_processor/src/lib.rs

```diff
@@ -762,205 +456,139 @@ impl<T: BeaconChainTypes> BeaconProcessorSend<T> {
     },
+    GossipBlock(AsyncFn),
     DelayedImportBlock {
-        peer_id: PeerId,
-        block: Box<GossipVerifiedBlock<T>>,
-        seen_timestamp: Duration,
-    },
-    GossipVoluntaryExit {
-        message_id: MessageId,
-        peer_id: PeerId,
-        voluntary_exit: Box<SignedVoluntaryExit>,
-    },
-    GossipProposerSlashing {
-        message_id: MessageId,
-        peer_id: PeerId,
-        proposer_slashing: Box<ProposerSlashing>,
-    },
-    GossipAttesterSlashing {
-        message_id: MessageId,
-        peer_id: PeerId,
-        attester_slashing: Box<AttesterSlashing<T::EthSpec>>,
-    },
-    GossipSyncSignature {
-        message_id: MessageId,
-        peer_id: PeerId,
-        sync_signature: Box<SyncCommitteeMessage>,
-        subnet_id: SyncSubnetId,
-        seen_timestamp: Duration,
-    },
-    GossipSyncContribution {
-        message_id: MessageId,
-        peer_id: PeerId,
-        sync_contribution: Box<SignedContributionAndProof<T::EthSpec>>,
-        seen_timestamp: Duration,
-    },
-    GossipLightClientFinalityUpdate {
-        message_id: MessageId,
-        peer_id: PeerId,
-        light_client_finality_update: Box<LightClientFinalityUpdate<T::EthSpec>>,
-        seen_timestamp: Duration,
-    },
-    GossipLightClientOptimisticUpdate {
-        message_id: MessageId,
-        peer_id: PeerId,
-        light_client_optimistic_update: Box<LightClientOptimisticUpdate<T::EthSpec>>,
-        seen_timestamp: Duration,
+        beacon_block_slot: Slot,
+        beacon_block_root: Hash256,
+        process_fn: AsyncFn,
     },
+    GossipVoluntaryExit(BlockingFn),
+    GossipProposerSlashing(BlockingFn),
+    GossipAttesterSlashing(BlockingFn),
+    GossipSyncSignature(BlockingFn),
+    GossipSyncContribution(BlockingFn),
+    GossipLightClientFinalityUpdate(BlockingFn),
+    GossipLightClientOptimisticUpdate(BlockingFn),
     RpcBlock {
-        block_root: Hash256,
-        block: Arc<SignedBeaconBlock<T::EthSpec>>,
-        seen_timestamp: Duration,
-        process_type: BlockProcessType,
-        should_process: bool,
+        process_fn: AsyncFn,
     },
-    ChainSegment {
-        process_id: ChainSegmentProcessId,
-        blocks: Vec<Arc<SignedBeaconBlock<T::EthSpec>>>,
-    },
-    Status {
-        peer_id: PeerId,
-        message: StatusMessage,
-    },
-    BlocksByRangeRequest {
-        peer_id: PeerId,
-        request_id: PeerRequestId,
-        request: BlocksByRangeRequest,
-    },
-    BlocksByRootsRequest {
-        peer_id: PeerId,
-        request_id: PeerRequestId,
-        request: BlocksByRootRequest,
-    },
-    GossipBlsToExecutionChange {
-        message_id: MessageId,
-        peer_id: PeerId,
-        bls_to_execution_change: Box<SignedBlsToExecutionChange>,
-    },
-    LightClientBootstrapRequest {
-        peer_id: PeerId,
-        request_id: PeerRequestId,
-        request: LightClientBootstrapRequest,
+    IgnoredRpcBlock {
```

Comment:
The `IgnoredRpcBlock` is a new variant. Previously we had a `should_process` field on the `RpcBlock` variant, we've removed that and the `IgnoredRpcBlock` variant is equivalent to `should_process == false`.

This was a necessary change since the `should_process` field is toggled inside the `work_reprocessing_queue` and it's hard to pass that value back into a future that was defined prior to us deciding that the work should be put in the reprocessing queue.

In effect, the `work_reprocessing_queue` will just map an `RpcBlock` into an `IgnoredRpcBlock` when it used to set `should_process` to `false`.